File: web/src/lib/cli/report.ts
import type { RunReport, StepReport } from '@/lib/db/cli';
import type { TestResultRow } from '@/lib/db/schema';
import type { MovedUnit } from '@/lib/model';

/**
 * Reading a step off the wire, shared by the two endpoints that take one.
 *
 * A completion sends every step at once and settles the run; the preview sends them
 * as they finish and settles nothing. The step itself is the same shape either way,
 * and it is worth it being literally the same code: the two differing on what a
 * `status` may be, or on how a body is bounded, would show up as a run whose steps
 * changed when it finished.
 *
 * The enums are checked here rather than left to Postgres because a rejected enum
 * value is a 500 by the time it is a constraint violation, and there is a machine on
 * the other end of the wire that needs telling what it got wrong.
 */

/**
 * More steps than this is a runner bug or a machine that has lost its mind, not a
 * test plan. Plans are drafted in handfuls; the largest thing here by far is the
 * bodies, and five hundred of those is already a request worth refusing.
 */
export const MAX_STEPS = 500;

/**
 * Ledger rows for one request, steps and run together. A unit only appears when it
 * moved, so a run over the hotel project's 45 tables reports a handful; this is the
 * ceiling that says a machine sending thousands has a bug rather than a finding.
 */
export const MAX_MOVED = 2_000;

/**
 * Per-body ceiling. Over it, the body is replaced by a marker rather than the request
 * being refused: a step's status, timing and assertions are what the report is for,
 * and losing all of that because one response was a 4MB file dump would be the wrong
 * trade. The run report reads bodies now, so the marker is rendered as the sentence it
 * is -- "too big to keep, 4.2MB came back" -- rather than as a mystery empty panel.
 */
const MAX_BODY_BYTES = 64 * 1024;

const STEP_STATUSES = new Set(['pending', 'passed', 'failed', 'skipped', 'error']);

/* Absent is `http`, not a rejection: a CLI older than the two other kinds omits the
   field, and every step it ever ran was a request. */
const STEP_KINDS = new Set(['http', 'sql', 'shell']);

/**
 * What a shell step printed, bounded the way `errorMessage` is. A build log is not a
 * step result, and the runner has already cut this to its last few KB — the end being
 * the half worth keeping, since a command that fails says why at the bottom.
 */
const MAX_OUTPUT = 16_000;

/** How many ledger rows a report is asking to write, which is what MAX_MOVED bounds. */
export function ledgerRows(steps: StepReport[], runMoved: MovedUnit[]): number {
  return runMoved.length + steps.reduce((n, s) => n + (s.moved?.length ?? 0), 0);
}

/** Null for anything that is not a step, which the caller turns into a 400. */
export function parseStep(entry: unknown): StepReport | null {
  if (!entry || typeof entry !== 'object') return null;
  const step = entry as Record<string, unknown>;

  // Both columns are NOT NULL and both are varchar(255), so an over-long value has to
  // be refused rather than trimmed: a truncated step id names a different step.
  const stepId = typeof step.stepId === 'string' ? step.stepId.trim() : '';
  const stepName = typeof step.stepName === 'string' ? step.stepName.trim() : '';
  if (!stepId || stepId.length > 255) return null;
  if (!stepName || stepName.length > 255) return null;

  if (typeof step.status !== 'string' || !STEP_STATUSES.has(step.status)) return null;

  // Same reasoning as the ids: `request_method` is varchar(10), and a truncated method
  // is a different method.
  const method = typeof step.method === 'string' ? step.method.trim().toUpperCase() : null;
  if (method !== null && (method.length === 0 || method.length > 10)) return null;

  const kind = step.kind === undefined || step.kind === null ? 'http' : step.kind;
  if (typeof kind !== 'string' || !STEP_KINDS.has(kind)) return null;

  const moved = parseMoved(step.moved);
  if (!moved) return null;

  return {
    stepId,
    stepName,
    status: step.status as TestResultRow['status'],
    kind: kind as TestResultRow['stepKind'],
    method,
    /* Held to null for the other two kinds rather than trusted. `route_pattern` is
       what the coverage grid counts, and a statement landing in it would draw a
       square for an endpoint named SELECT. The runner already leaves it empty for a
       sql or shell step; this is the half that does not depend on it being right. */
    routePattern: kind === 'http' ? str(step.routePattern, 2048) : null,
    /* The URL for a request, the statement or the command for the other two, with
       the variables filled in and the secrets masked. One column because it answers
       one question -- what this step actually did. */
    requestUrl: str(step.requestUrl, 4096),
    requestBody: bounded(step.requestBody),
    responseStatus: int(step.responseStatus),
    responseBody: bounded(step.responseBody),
    responseTimeMs: int(step.responseTimeMs),
    /* `signed`, not `int`: a fixture that deletes moves a count down, and flooring
       that at zero erases the finding. Absent stays absent rather than becoming 0,
       because 0 rows is the answer a verification step exists to catch -- "returned
       201, wrote nothing" -- and it must be tellable from a step with no rows at
       all. Same for an exit code, where 0 is success. */
    rowCount: step.rowCount === undefined ? null : signed(step.rowCount),
    exitCode: step.exitCode === undefined ? null : signed(step.exitCode),
    output: str(step.output, MAX_OUTPUT),
    assertions: bounded(step.assertions),
    errorMessage: str(step.errorMessage, 4000),
    moved,
  };
}

/**
 * What moved, as the runner measured it. Null for anything that is not a ledger.
 *
 * A unit names a table or a watched path, so an over-long one is refused rather than
 * trimmed for the same reason a step id is: a truncated name is a different unit.
 */
export function parseMoved(value: unknown): MovedUnit[] | null {
  if (value === undefined || value === null) return [];
  if (!Array.isArray(value)) return null;

  const out: MovedUnit[] = [];
  for (const entry of value) {
    if (!entry || typeof entry !== 'object') return null;
    const unit = entry as Record<string, unknown>;

    const name = typeof unit.unit === 'string' ? unit.unit.trim() : '';
    if (!name || name.length > 1024) return null;

    const rows = signed(unit.rows);
    if (rows === null) return null;

    out.push({
      unit: name,
      rows,
      from: str(unit.from, 1024),
      to: str(unit.to, 1024),
    });
  }
  return out;
}

/** Optional, self-reported, and length-bounded because the columns are. */
export function str(value: unknown, max: number): string | null {
  if (typeof value !== 'string') return null;
  const clean = value.trim();
  return clean ? clean.slice(0, max) : null;
}

/** Whole, non-negative, and inside the INTEGER the column is. */
export function int(value: unknown): number | null {
  if (typeof value !== 'number' || !Number.isFinite(value)) return null;
  const whole = Math.trunc(value);
  if (whole < 0) return null;
  return Math.min(whole, 2_147_483_647);
}

/**
 * Whole and signed, unlike `int`: a delete moving a count down is as much a finding as
 * an insert, and flooring it at zero would erase the half of the ledger that is
 * hardest to get any other way. `rows_moved` is a bigint, so the bound is what a JS
 * number can carry exactly.
 */
export function signed(value: unknown): number | null {
  if (typeof value !== 'number' || !Number.isFinite(value)) return null;
  const whole = Math.trunc(value);
  return Number.isSafeInteger(whole) ? whole : null;
}

/**
 * A JSONB value, or a note saying how big the one we were sent was.
 *
 * `undefined` becomes null rather than being passed through: `jsonb` has no
 * representation for it, and a column holding the JSON literal `null` and a column
 * holding SQL NULL are two different answers to "was there a body".
 */
export function bounded(value: unknown): unknown {
  if (value === undefined || value === null) return null;
  // Safe to stringify unconditionally: this came out of `JSON.parse`, so there is
  // nothing in it that cannot go back the other way.
  const bytes = Buffer.byteLength(JSON.stringify(value) ?? 'null', 'utf8');
  return bytes > MAX_BODY_BYTES ? { truncated: true, bytes } : value;
}

const OUTCOMES = new Set(['passed', 'failed', 'error']);

export type ParsedReport =
  | { ok: true; instanceId: string; report: RunReport }
  | { ok: false; error: string };

/**
 * A finished run, off the wire.
 *
 * Shared by the two routes that take one -- `jobs/[id]/complete`, which settles a run
 * somebody queued, and `executions`, which records one nobody did. Literally the same
 * code rather than the same shape twice, for the reason `parseStep` is: the two
 * differing on what an outcome may be, or on whether a `failed` run must contain a
 * failed step, would mean a run reported from a terminal and a run reported from the
 * queue could disagree about the same facts.
 *
 * `instanceId` is read here too, though only one of the two routes authorises with
 * it. Both use it as a sighting -- a machine that finishes a long run and then goes
 * idle should not read as disconnected in the gap before its next poll.
 */
export function parseReport(body: unknown): ParsedReport {
  if (!body || typeof body !== 'object' || Array.isArray(body)) {
    return { ok: false, error: 'invalid_body' };
  }
  const input = body as Record<string, unknown>;

  const instanceId = typeof input.instanceId === 'string' ? input.instanceId.trim() : '';
  if (!instanceId || instanceId.length > 255) return { ok: false, error: 'invalid_instance' };

  const outcome = input.outcome;
  if (typeof outcome !== 'string' || !OUTCOMES.has(outcome)) {
    return { ok: false, error: 'invalid_outcome' };
  }

  const rawSteps = input.steps ?? [];
  if (!Array.isArray(rawSteps)) return { ok: false, error: 'invalid_steps' };
  if (rawSteps.length > MAX_STEPS) return { ok: false, error: 'too_many_steps' };

  const steps: StepReport[] = [];
  for (const entry of rawSteps) {
    const step = parseStep(entry);
    if (!step) return { ok: false, error: 'invalid_step' };
    steps.push(step);
  }

  /* The verdict and the steps have to be able to agree, in both directions.
     A `passed` run with a step that did not pass is the obvious half. The other half
     is a `failed` run in which nothing failed, and it is the one that bites: the runs
     table renders that status as "A step failed", the run report reads its headline
     off the cells and finds nothing stopped there, and the two disagree on screen
     about the same row.
     Refused rather than corrected, because either half could be the true one and
     guessing wrong is how that disagreement gets written down as fact. `skipped` is
     allowed through a passing run -- a step nobody reached is not a step that failed --
     and a run that failed before its first step is an `error`, which is what the run
     report already has copy for. */
  const broke = steps.some((s) => s.status === 'failed' || s.status === 'error');
  if ((outcome === 'passed' && broke) || (outcome === 'failed' && !broke)) {
    return { ok: false, error: 'contradictory_outcome' };
  }

  const moved = parseMoved(input.moved);
  if (!moved) return { ok: false, error: 'invalid_moved' };
  if (ledgerRows(steps, moved) > MAX_MOVED) return { ok: false, error: 'too_much_state' };

  return {
    ok: true,
    instanceId,
    report: {
      outcome: outcome as RunReport['outcome'],
      durationMs: int(input.durationMs),
      containerId: str(input.containerId, 64),
      errorMessage: str(input.errorMessage, 4000),
      steps,
      moved,
      stateNote: str(input.stateNote, 4000),
    },
  };
}

File: cli/internal/cloud/report.go
package cloud

import (
	"encoding/json"
	"fmt"
	"strings"
	"unicode/utf8"

	"github.com/gritqa/cli/internal/plan"
	"github.com/gritqa/cli/internal/run"
)

// The complete route's own limits. They are duplicated here on purpose: it
// refuses a whole report over one over-long field, and a run whose results were
// thrown away at the door is worse than one reported with a trimmed name.
const (
	maxSteps     = 500
	maxBody      = 64 << 10
	maxID        = 255
	maxMethod    = 10
	maxPattern   = 2048
	maxURL       = 4096
	maxMessage   = 4000
	maxContainer = 64
	maxOutput    = 16000
)

type Report struct {
	InstanceID   string  `json:"instanceId"`
	Outcome      string  `json:"outcome"`
	DurationMs   int64   `json:"durationMs"`
	ContainerID  string  `json:"containerId,omitempty"`
	ErrorMessage string  `json:"errorMessage,omitempty"`
	Steps        []Step  `json:"steps"`
	Moved        []Moved `json:"moved,omitempty"`
	StateNote    string  `json:"stateNote,omitempty"`
}

type Step struct {
	StepID   string `json:"stepId"`
	StepName string `json:"stepName"`
	Status   string `json:"status"`
	// Kind is omitted for an HTTP step, which the route reads as http, so an
	// older CLI reporting into a newer dashboard still says something true.
	Kind           string          `json:"kind,omitempty"`
	Method         string          `json:"method,omitempty"`
	RoutePattern   string          `json:"routePattern,omitempty"`
	RequestURL     string          `json:"requestUrl,omitempty"`
	RequestBody    json.RawMessage `json:"requestBody,omitempty"`
	ResponseStatus int             `json:"responseStatus,omitempty"`
	ResponseBody   json.RawMessage `json:"responseBody,omitempty"`
	ResponseTimeMs int64           `json:"responseTimeMs,omitempty"`
	// RowCount is a sql step's evidence: rows a query came back with, or rows a
	// fixture moved. Carried as a pointer because zero is the interesting answer
	// -- "returned 201, wrote nothing" is the bug this whole step type exists
	// for, and omitempty would drop exactly that.
	RowCount *int64 `json:"rowCount,omitempty"`
	// ExitCode is a shell step's, and zero is the ordinary answer, so it is a
	// pointer for the same reason. It is deliberately not folded into
	// ResponseStatus: an exit code of 0 must never render as HTTP 0.
	ExitCode *int `json:"exitCode,omitempty"`
	// Output is what a shell step printed, already masked by the runner.
	Output       string  `json:"output,omitempty"`
	Assertions   []Check `json:"assertions,omitempty"`
	ErrorMessage string  `json:"errorMessage,omitempty"`
	Moved        []Moved `json:"moved,omitempty"`
}

type Check struct {
	Type     string `json:"type"`
	Operator string `json:"operator"`
	Target   string `json:"target,omitempty"`
	Expected string `json:"expected,omitempty"`
	Actual   string `json:"actual,omitempty"`
	Passed   bool   `json:"passed"`
	Found    bool   `json:"found"`
}

// Moved is one unit that changed, as the ledger records it.
type Moved struct {
	Unit string `json:"unit"`
	Rows int64  `json:"rows"`
	From string `json:"from,omitempty"`
	To   string `json:"to,omitempty"`
}

// Failed is the report for a job this machine took and could not carry out — a
// sandbox that would not boot, a payload it could not read. The run is settled as
// an error with the reason attached, which is what the dashboard has copy for.
func Failed(instanceID, why string) Report {
	return Report{InstanceID: instanceID, Outcome: "error", ErrorMessage: cut(why, maxMessage)}
}

// Reported turns a finished run into the body the complete route accepts.
//
// The plan comes in beside the result because two of the columns are the plan's
// answer rather than the engine's: routePattern is the URL as written, which is
// what stays comparable across runs, and requestBody is what the plan declared.
func Reported(instanceID string, p *plan.Plan, base string, res *run.Result, container string) Report {
	declared := map[string]plan.Step{}
	if p != nil {
		for _, s := range p.Steps {
			declared[s.ID] = s
		}
	}

	steps := make([]Step, 0, len(res.Steps))
	for i, s := range res.Steps {
		steps = append(steps, step(i, s, declared[s.ID], base))
	}

	outcome := string(res.Status)
	kept, dropped := fit(steps, outcome)

	rep := Report{
		InstanceID:  instanceID,
		Outcome:     outcome,
		DurationMs:  res.Elapsed.Milliseconds(),
		ContainerID: cut(container, maxContainer),
		Steps:       kept,
		Moved:       units(res.Moved),
		StateNote:   cut(res.StateErr, maxMessage),
	}

	var notes []string
	if r := why(steps); r != "" {
		notes = append(notes, r)
	}
	if dropped > 0 {
		notes = append(notes, fmt.Sprintf("this run had %d steps and %d of them are not reported here",
			len(steps), dropped))
	}
	rep.ErrorMessage = cut(strings.Join(notes, " — "), maxMessage)
	return rep
}

// why is the reason a run did not pass, read off the first step that did not. The
// step rows carry the same thing, but they are what fit() may have dropped, and a
// run whose error_message is empty reads on the dashboard as a failure with no
// cause -- which is what sent the user here to ask why.
func why(steps []Step) string {
	first, others := "", 0
	for _, s := range steps {
		r := reason(s)
		if r == "" {
			continue
		}
		if first == "" {
			first = s.StepName + ": " + r
			continue
		}
		others++
	}
	if others > 0 {
		return fmt.Sprintf("%s (and %d more steps did not pass)", first, others)
	}
	return first
}

func reason(s Step) string {
	switch s.Status {
	case string(run.StepError), string(run.StepSkipped):
		return s.ErrorMessage
	case string(run.StepFailed):
		var out []string
		for _, c := range s.Assertions {
			if !c.Passed {
				out = append(out, unmet(c))
			}
		}
		if len(out) == 0 {
			return s.ErrorMessage
		}
		return strings.Join(out, "; ")
	}
	return ""
}

// unmet says what one assertion wanted and what it got. Deliberately the wire's
// own words -- type, operator, expected -- rather than the operator prose the
// terminal prints: that map is presentation and lives with the renderer.
func unmet(c Check) string {
	target := c.Target
	if target == "" {
		target = c.Type
	}
	if c.Operator == string(plan.Exists) {
		return "expected " + target + " to be there, and it was not"
	}
	got := c.Actual
	if got == "" {
		got = "nothing"
	}
	return fmt.Sprintf("expected %s %s %s, got %s", target, c.Operator, c.Expected, got)
}

func step(i int, s run.StepResult, declared plan.Step, base string) Step {
	id := cut(s.ID, maxID)
	if id == "" {
		id = fmt.Sprintf("step-%d", i+1)
	}
	name := cut(s.Name, maxID)
	if name == "" {
		name = id
	}

	out := Step{
		StepID:   id,
		StepName: name,
		Status:   string(s.Status),
		Kind:     kindOf(s),
		Method:   cut(strings.ToUpper(s.Method), maxMethod),
		// RequestURL is the interpolated statement or command for a non-HTTP
		// step, which is the same promise it makes for an HTTP one: what
		// actually went out, replayable.
		RoutePattern:   cut(pattern(declared, s, base), maxPattern),
		RequestURL:     cut(s.URL, maxURL),
		ResponseStatus: s.Code,
		ResponseBody:   payload(s.Body),
		ResponseTimeMs: s.Elapsed.Milliseconds(),
		ErrorMessage:   cut(s.Err, maxMessage),
		Moved:          units(s.Moved),
	}
	switch s.Kind {
	case plan.SQLStep:
		rows := s.RowsAffected
		out.RowCount = &rows
	case plan.ShellStep:
		code := s.ExitCode
		out.ExitCode = &code
		out.Output = cut(s.Stdout, maxOutput)
	}
	if len(declared.Request.Body) > 0 {
		if b, err := json.Marshal(declared.Request.Body); err == nil {
			out.RequestBody = payload(b)
		}
	}
	for _, c := range s.Checks {
		out.Assertions = append(out.Assertions, Check{
			Type: string(c.Type), Operator: string(c.Operator), Target: c.Target,
			Expected: c.Expected, Actual: c.Actual, Passed: c.Passed, Found: c.Found,
		})
	}
	return out
}

// kindOf spells a step's kind for the wire, leaving http empty so the common case
// costs nothing and an older dashboard reads it the way it always has.
func kindOf(s run.StepResult) string {
	if s.IsHTTP() {
		return ""
	}
	return string(s.Kind)
}

// pattern is the URL as the plan wrote it, placeholders and all. The step's own
// URL is the concrete one, so keeping both is what makes "every run that touched
// this route" answerable later.
//
// Empty for a non-HTTP step, and it has to be: route_pattern is what the coverage
// grid counts, and a SQL statement landing in that column would invent an endpoint
// named SELECT. A sql step proving a row was written is evidence about an endpoint
// but it is not traffic to one.
func pattern(declared plan.Step, s run.StepResult, base string) string {
	if !s.IsHTTP() {
		return ""
	}
	if declared.Request.URL != "" {
		return strings.TrimPrefix(declared.Request.URL, strings.TrimRight(base, "/"))
	}
	return strings.TrimPrefix(s.URL, strings.TrimRight(base, "/"))
}

// payload keeps a body renderable as jsonb: parsed when it is JSON, quoted when it
// is not, and replaced by its own size when it is too big to be worth carrying.
// The route does the same swap; doing it here as well keeps the request small.
func payload(b []byte) json.RawMessage {
	if len(b) == 0 {
		return nil
	}
	if len(b) > maxBody {
		return json.RawMessage(fmt.Sprintf(`{"truncated":true,"bytes":%d}`, len(b)))
	}
	if json.Valid(b) {
		return json.RawMessage(b)
	}
	quoted, err := json.Marshal(string(b))
	if err != nil {
		return nil
	}
	return quoted
}

func units(ms []run.Moved) []Moved {
	out := make([]Moved, 0, len(ms))
	for _, m := range ms {
		out = append(out, Moved{Unit: m.Unit, Rows: m.Rows, From: m.From, To: m.To})
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// fit trims a report to the route's ceiling. A failed run has to keep a step that
// failed or the route refuses the whole thing for contradicting itself — and the
// step that broke is the one worth keeping anyway.
func fit(steps []Step, outcome string) ([]Step, int) {
	if len(steps) <= maxSteps {
		return steps, 0
	}
	kept := make([]Step, maxSteps)
	copy(kept, steps[:maxSteps])
	dropped := len(steps) - maxSteps

	if outcome != "failed" || broke(kept) {
		return kept, dropped
	}
	for _, s := range steps[maxSteps:] {
		if s.Status == string(run.StepFailed) || s.Status == string(run.StepError) {
			kept[len(kept)-1] = s
			return kept, dropped
		}
	}
	return kept, dropped
}

func broke(steps []Step) bool {
	for _, s := range steps {
		if s.Status == string(run.StepFailed) || s.Status == string(run.StepError) {
			return true
		}
	}
	return false
}

// cut trims to a byte ceiling without leaving half a rune behind: Postgres text
// refuses invalid UTF-8, so a truncated accented message would cost the whole run.
func cut(s string, n int) string {
	s = strings.TrimSpace(s)
	if len(s) <= n {
		return s
	}
	s = s[:n]
	for len(s) > 0 && !utf8.ValidString(s) {
		s = s[:len(s)-1]
	}
	return s
}
